test/e2e/storage/pvc_protection.go

```diff
@@ -0,0 +1,125 @@
+		pvcCreatedAndNotDeleted = true
+
+		By("Waiting for PVC to become Bound")
+		err = framework.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, client, nameSpace, pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
+		Expect(err).NotTo(HaveOccurred(), "Failed waiting for PVC to be bound %v", err)
+
+		By("Checking that PVC Protection finalizer is set")
+		pvc, err = client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(pvc.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred(), "While getting PVC status")
+		Expect(slice.ContainsString(pvc.ObjectMeta.Finalizers, volumeutil.PVCProtectionFinalizer, nil)).To(BeTrue())
+	})
+
+	AfterEach(func() {
+		if pvcCreatedAndNotDeleted {
+			framework.DeletePersistentVolumeClaim(client, pvc.Name, nameSpace)
+		}
+	})
+
+	It("Verify \"immediate\" deletion of a PVC that is not in active use by a pod", func() {
+		By("Deleting the PVC")
+		err = client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(pvc.Name, metav1.NewDeleteOptions(0))
+		Expect(err).NotTo(HaveOccurred(), "Error deleting PVC")
+		waitForPersistentVolumeClaimBeRemoved(client, pvc.Namespace, pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
+		pvcCreatedAndNotDeleted = false
+	})
+
+	It("Verify that PVC in active use by a pod is not removed immediatelly", func() {
+		By("Creating a Pod that becomes Running and therefore is actively using the PVC")
+		pvcClaims := []*v1.PersistentVolumeClaim{pvc}
+		pod, err := framework.CreatePod(client, nameSpace, nil, pvcClaims, false, "")
+		Expect(err).NotTo(HaveOccurred(), "While creating pod that uses the PVC or waiting for the Pod to become Running")
+
+		By("Deleting the PVC, however, the PVC must not be removed from the system as it's in active use by a pod")
+		err = client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(pvc.Name, metav1.NewDeleteOptions(0))
+		Expect(err).NotTo(HaveOccurred(), "Error deleting PVC")
+
+		By("Checking that the PVC status is Terminating")
+		pvc, err = client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(pvc.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred(), "While checking PVC status")
+		Expect(pvc.ObjectMeta.DeletionTimestamp).NotTo(Equal(nil))
+
+		By("Deleting the pod that uses the PVC")
+		err = framework.DeletePodWithWait(f, client, pod)
+		Expect(err).NotTo(HaveOccurred(), "Error terminating and deleting pod")
+
+		By("Checking that the PVC is automativally removed from the system because it's no longer in active use by a pod")
+		waitForPersistentVolumeClaimBeRemoved(client, pvc.Namespace, pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
+		pvcCreatedAndNotDeleted = false
+	})
+})
+
+// waitForPersistentVolumeClaimBeRemoved waits for a PersistentVolumeClaim to be removed from the system until timeout occurs, whichever comes first.
+func waitForPersistentVolumeClaimBeRemoved(c clientset.Interface, ns string, pvcName string, Poll, timeout time.Duration) error {
```

Comment:
It would be nice to move this to `test/e2e/framework/util.go`, maybe in the aforementioned PR with scheduler tests.

Reply:
Should private func that is used only on one place be made public?

Reply:
It won't hurt. I like them all together so I can see what `WaitFor<something>` are available when I write my test. It's harder to find when it's just in `pvc_protection.go`.